src/test/scala/freeeventsourcing/ProcessTestSupport.scala

```diff
@@ -0,0 +1,140 @@
+        }
+        .flatMap(checkResult)
+      new MatchResult(
+        y.isRight,
+        y.fold(identity, _ ⇒ "?"),
+        s"matched the process expectations but should not have"
+      )
+    }
+  }
+
+  type Action[+A] = ProcessAction[BC, A]
+  type M[A] = ProcessMonad[BC, A]
+
+  sealed trait Expectation
+  object Expectation {
+    case class ExpectAwaitEvent[S <: WithEventType](selector: S, result: S#Event) extends Expectation
+    case class ExpectWaitUntil(instant: Instant) extends Expectation
+    case class ExpectCommand[A <: Aggregate, Cmd <: A#Command](
+      aggregateType: A, aggregate: A#Id, command: A#Command, result: Cmd#Error Xor Unit
+    ) extends Expectation
+    case object ExpectEnd extends Expectation
+  }
+
+  private[this] object Implementation {
+    import Expectation._
+
+    type OrFail[A] = Xor[String, A]
+    type Expectations[A] = StateT[OrFail, List[Expectation], A]
+    object Expectations {
+      def next(received: String): Expectations[Expectation] = lift {
+        case h :: t ⇒ Xor.right((t, h))
+        case Nil    ⇒ Xor.left(s"No more expectations, but got ${received}")
+      }
+
+      def check[A](action: Action[A]): Expectations[A] = next(action.toString).flatMap { exp ⇒
+        compare.lift((action, exp)).getOrElse {
+          lifted(Xor.left(s"Expected ${exp} but got ${action}"))
+        }
+      }
+
+      private[this] def compare[A]: PartialFunction[(Action[A], Expectation), Expectations[A]] = {
+        case (AwaitEvent(s1), ExpectAwaitEvent(s2, result)) if s1 == s2 ⇒
+          ok(result)
+        case (AwaitEvent(s1), ExpectAwaitEvent(s2, _)) ⇒
+          fail(s"AwaitEvent: selector mismatch ${s1} != ${s2}")
+
+        case (WaitUntil(i1), ExpectWaitUntil(i2)) if i1 == i2 ⇒
+          ok(())
+        case (WaitUntil(i1), ExpectWaitUntil(i2)) ⇒
+          fail(s"WaitUntil: time is different: ${i1} != ${i2}")
+
+        case (End(), ExpectEnd) ⇒
+          ok(())
+
+        case (Execute(at1, _, _, _), ExpectCommand(at2, _, _, _)) if at1 != at2 ⇒
+          fail(s"Execute: Aggregate type is different: ${at1} != ${at2}")
+        case (Execute(_, a1, _, _), ExpectCommand(_, a2, _, _)) if a1 != a2 ⇒
+          fail(s"Execute: Aggregate id is different: ${a1} != ${a2}")
+        case (Execute(_, _, c1, _), ExpectCommand(_, _, c2, _)) if c1 != c2 ⇒
+          fail(s"Execute: Command is different: ${c1} != ${c2}")
+        case (ex @ Execute(_, _, _, errorHandler), e @ ExpectCommand(_, _, _, r)) ⇒
+          r.fold({ error ⇒
+            //could be proved, but for tests a ClassCastException is ok
+            val err: ex.Error = error.asInstanceOf[ex.Error]
+            val subprocess: M[Unit] = errorHandler(err).asInstanceOf[M[Unit]]
+            val x = subprocess.foldMap(Implementation.Transform)
+            lift(s ⇒ x.run(s))
+          }, ok)
+
+        // TODO case FirstOf(alternatives) ⇒ ???
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Take the required action to fix the issue indicated by this comment.](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2580073129/issues/source?bid=3298199&fileBranchId=3298202#l124)